lib/Model/EventTree.js

```diff
@@ -0,0 +1,114 @@
+module.exports = EventTree;
+
+function EventTree(parent, segment) {
+  this.parent = parent;
+  this.segment = segment;
+  this.children = null;
+  this.listeners = null;
+}
+
+EventTree.prototype.destroy = function() {
+  // Ignore calls to destroy a root node
+  if (!this.parent) return;
+
+  // Remove reference this node from its parent
+  if (hasOtherKeys(this.parent.children, this.segment)) {
+    delete this.parent.children[this.segment];
+    return;
+  }
+  this.parent.children = null;
+
+  // Destroy parent if it no longer has any dependents
+  if (!this.parent.listeners) {
+    this.parent.destroy();
+  }
+};
+
+EventTree.prototype.getChild = function(segments) {
+  var node = this;
+  for (var i = 0; i < segments.length; i++) {
+    var segment = segments[i];
+    node = node.children && node.children[segment];
+    if (!node) return;
+  }
+  return node;
+};
+
+EventTree.prototype.getOrCreateChild = function(segments) {
+  var node = this;
+  for (var i = 0; i < segments.length; i++) {
+    var segment = segments[i];
+    if (!node.children) {
+      node.children = {};
+    }
+    var node = node.children[segment] ||
+      (node.children[segment] = new EventTree(node, segment));
+  }
+  return node;
+};
+
+EventTree.prototype.addListener = function(segments, listener) {
+  var node = this.getOrCreateChild(segments);
+  if (!node.listeners) {
+    node.listeners = [listener];
+    return;
+  }
+  var i = node.listeners.indexOf(listener);
```

Comment:
These `addListener` and `removeListener` implementations are O(n) with the number of listeners on the current node, so a rendering path that adds/removes n reactive functions based on the same input or output path would be roughly (n^2)/2.

If we supported ES6, we could use a `Set` for the listeners. On ES5, one alternative would be to attach a unique id to each `Fn` instance and use that to key an object of listeners here.

Thoughts?

Reply:
I'm not sure what is the most performant, because there may be a tradeoff between adding/removing efficiency and iteration efficiency. I expect that iterating through an array may be more efficient than the keys of an object, and I'd like the implementation to guard against adding the same value twice to the same path, so we'd have to do some sort of equality check with an object-based map, and I think that would be the same O(n).

There is a good chance that Set is more efficient than Array, but we'd probably have to implement it both ways and test it in practice.

Reply:
I tried using a Set, and the overhead of creating a Set instance vs. an Array made it slower in practice. Sticking with arrays and the index check for now.